[XLA:CPU][oneDNN] Pad MemrefInfo pod

akhilgoe opened the pull request:
This PR pads the MemrefInfoPOD emitted on stack so that it aligns with the system cacheline.

penpornk commented on xla/service/cpu/onednn_memory_util.cc:
```diff
@@ -45,9 +45,10 @@ namespace cpu {
 struct MemrefInfoPOD {
   int64_t dtype;
   int64_t rank;
+  void* data;
```
What cache line size are we aligning for? Also, is there a reason for moving `data` up? (We can still pad when it's the last member.)

akhilgoe replied:
Aligning for 64 bytes, will add a comment. 
Here's my thought process on moving `data` up:
oneDNN restricts tensor dimensions to a maximum of 12, though in practice we rarely use dimensions larger than 3 or 4.  By moving `data` up, I think we can reduce the total number of cacheline loads from 3 to 2 as dtype, rank, data and dimensions can fit within a single cacheline, while strides can fit within a second cacheline.
Please let me know if this makes sense or if there is something that I missed.

penpornk commented on xla/service/cpu/onednn_memory_util.cc:
```diff
@@ -45,9 +45,10 @@ namespace cpu {
   int64_t dtype;
   int64_t rank;
+  void* data;
+  int64_t pad;
```
`unused` might be a better name. `pad` could be misinterpret as some padding setting for buffer.  
```suggestion
  int64_t unused;
```
Please also add a comment after this variable explaining that it's just to pad the struct to fit the cache line of size X bytes.

akhilgoe replied:
Done

penpornk commented on xla/service/cpu/tests/onednn_memory_util_test.cc:
```diff
@@ -0,0 +1,109 @@
+                           std::tuple<PrimitiveType, std::vector<int64_t>>> {
+ protected:
+  constexpr static const char* pad_test_pattern_ = R"(
+    CHECK: %[[mref0:[0-9]+]] = insertvalue
+    CHECK: %[[mref1:[0-9]+]] = insertvalue
+    CHECK-SAME: [[arr:\[12 x i64\]]], [[arr]] } %[[mref0]], i64 255, 3
```
Nit: We don't need to capture `mref1` and `arr`. 
```suggestion
    CHECK: %[[mref0:[0-9]+]] = insertvalue
    CHECK: %{{[0-9]+}} = insertvalue
    CHECK-SAME: [12 x i64], [12 x i64] } %[[mref0]], i64 255, 3
```

akhilgoe replied:
Modified the filecheck pattern to match the new tests.

penpornk commented on xla/service/cpu/tests/onednn_memory_util_test.cc:
```diff
@@ -0,0 +1,109 @@
+
+INSTANTIATE_TEST_SUITE_P(
+    MemoryUtilTestSuite, MemoryUtilTest,
+    ::testing::Combine(::testing::ValuesIn({S8, S16, BF16, F16, F32}),
+                       ::testing::Values(std::vector<int64_t>({30}),
+                                         std::vector<int64_t>({30, 40}),
+                                         std::vector<int64_t>({30, 40, 50}))),
```
I don't think we need to input many dtypes/shapes since we don't check for them in the filecheck. (Or we could parameterize on just the shape, and add a check line to look for the actual sizes.)

penpornk reviewed (changes requested):
Thank you for the refactor!

akhilgoe commented:
Hi @penpornk, I have added a commit to address your comments. Please take a look whenever possible. Thanks!

penpornk reviewed (changes requested):
Thank you for the changes! Could you please add a microbenchmark in a test file that tracks the performance? Please also post the results.

See examples in:
https://github.com/openxla/xla/tree/main/xla/service/cpu/benchmarks

penpornk commented on xla/backends/cpu/benchmarks/onednn_util_benchmark_test.cc:
```diff
@@ -0,0 +1,95 @@
+/* Copyright 2024 The OpenXLA Authors.
```
```suggestion
/* Copyright 2025 The OpenXLA Authors.
```

penpornk commented on xla/backends/cpu/benchmarks/onednn_util_benchmark_test.cc:
```diff
@@ -0,0 +1,95 @@
+
+namespace xla::cpu {
+
+static void BM_oneDNN_MM(benchmark::State& state) {
```
Since this is just a matmul benchmark and doesn't have anything util-specific, I think the file name should just be onednn_matmul_benchmark_test.cc.

penpornk commented on xla/backends/cpu/benchmarks/hlo_benchmark_runner.h:
```diff
@@ -41,7 +41,8 @@ absl::Status RunHloBenchmark(benchmark::State& state,
                              absl::Span<const Literal* const> args,
                              StrToStrMapping replacements = {},
-                             bool disable_parallel_task_assigner = false);
+                             bool disable_parallel_task_assigner = false,
+                             bool use_thunk_runtime = true);
```
The function interface has changed. Parameters are now put in a struct `HloBenchmarkOptions`. Please resolve conflicts and add the `use_thunk_runtime` option to the struct instead.

penpornk reviewed (changes requested):
Thank you for the changes!
Just curious --  did you see any significant performance difference from padding when running this newly added matmul microbenchmarks? If so, please post some results. If not, please reply that you didn't see significant differences.

akhilgoe commented:
Sorry, I missed posting the benchmark results earlier. I'll resolve the conflicts and post the benchmark results.

akhilgoe commented:
Please find below the micro-benchmark data measured on 2 SPR CPUs

| Benchmark case | DType | Time (ns.) (main branch) | Time (ns.) (this branch) |  | CPU (ns.) (main branch) | CPU (ns.) (this branch) | | Iterations (main branch) | Iterations (this branch) |
|      :---:      |     :---:      |     :---:      |     :---:      |      :---:      |     :---:      |     :---:      |     :---:      |     :---:      |     :---:      |
| 512       | F32         | 149330   | 149312    | | 293850   | 291759     | | 2380  | 2391  |
| 1024     | F32         | 286717   | 280965    | | 559425   | 559133     | | 1240  | 1247  |
| 2048     | F32         | 571721   | 569529    | | 1132277 | 1132702   | | 580    | 594    |
| 512       | BF16       | 51022     | 49534      | | 96314     | 93456       | | 7571  | 7739  |
| 1024     | BF16       |  72501    | 71314      | | 140757   | 138042     | | 4988  | 5007  |
| 2048     | BF16       | 130331   | 131698    | | 257784   | 259867     | | 2590  | 2652  |

penpornk commented on xla/backends/cpu/benchmarks/onednn_matmul_benchmark_test.cc:
```diff
@@ -0,0 +1,97 @@
+
+namespace xla::cpu {
+
+static void BM_oneDNN_MM(benchmark::State& state) {
```
We probably should merge this with `dot_benchmark_test` in the future. But I'm fine with having this file until oneDNN thunks are properly enabled in the new runtime. (No action needed.)

penpornk reviewed (changes requested):
Thank you for the changes and sorry for the delay! One last nit please.

akhilgoe commented:
Thanks @penpornk for the review! I addressed your comment. I also removed the `ENABLE_ONEDNN_V3` conditional directive from the new files because I saw that it has been removed elsewhere.

penpornk reviewed (approved):
Thank you very much for the changes!

akhilgoe commented:
Hi @penpornk, it seems [this commit](https://github.com/openxla/xla/commit/a371e2252b9427ec22ea1316639e1eeb3482d0d3#diff-d73463f787ab614ef15deaeef00ab47a6c82ede363a39965af617a0e5701c5aa) from this morning deprecated and removed some files that the tests in this PR depended on. I have updated the file paths. Please take a look.
